New RPC feature : rollbackchain

ptschip opened the pull request:
This is a "hidden" feature in the same way as reconsiderblock or invalidateblock are also hidden features that do not show up in the "help" list of RPC calls.  Rolling back the chain can be useful in testing and also has real application for node operators that are trying to get back onto the correct chain either after a failed fork attempt;  the operator can roll the chain back before the fork point and then invalidate and reconsider the chains they wish to follow.  

This can also be useful in fork testing..when the tester can repeatedly roll the chain back to before the fork point and then try to join with the correct fork once again.

To use the feature simply to  "rollbackchain <chainheight to roll back to>"

ptschip committed:
Create rpc to roll back the blockchain to any height

This is a hidden feature which can be used during testing
or by users that are having difficulty syncing to the correct
chain and have tried reconsiderblock but have failed. By rolling
back beyond the fork point a user can then invalidate and reconsider
the correct chain.

ptschip committed:
Fix bug in reconsiderblock

We need to make sure to update the ui after we reconsider. This
was not always happening if we rolled the chain back and then reconsidered
a block again.

ptschip committed:
Remove unnecessary spaces

ptschip committed:
Limit how far a rollbackchain can go

In case of operator error, limit the rollback to 100 blocks
at a time.  An operator may for instance want to roll back
10 blocks and then enter "rollbackchain 10" which would have
the effect of rolling the chain all the way to block height 10
rather than just 10 blocks.

gandrewstone commented:
Have you tried rolling back 1000s of blocks?  I found a few years ago that doing so would fill up the mempool with the committed transactions in the unrolled blocks and these txns would eventually slow the performance of DisconnectTip() dramatically.  It would be good to solve this issue if we are adding RPC support for rollbacks.  I was thinking a simple solution is to clear the mempool and not add committed transactions to it if the rollback will be greater than 100 blocks (say).  These transactions are unlikely to be useful in large rollbacks anyway...

ptschip committed:
Do no resurrect transactions when doing rollbackchain

Resurrecting transactions is not really needed since the operator
may be following a different chaind with different txns and also
it slows down the process of rollback.

ptschip committed:
fix formatting
